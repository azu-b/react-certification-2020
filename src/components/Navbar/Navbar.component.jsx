import React, { useState, useRef } from 'react';
import { Link } from 'react-router-dom';
import useOutsideClick from '../../utils/hooks/useOutsideClick';
import SearchBar from '../SearchBar';
import { Container, Menu, MenuItem, OpenMenu, HamburgerButton } from './Navbar.styled';
import { HamburgerIcon } from '../../icons';

const Navbar = () => {
  const openMenuRef = useRef();
  const [isOpen, setOpen] = useState(false);

  const toggleMenu = () => {
    setOpen(!isOpen);
  };

  const menuItems = (
    <>
      <MenuItem>
        <Link to="/" onClick={toggleMenu}>
          Home
        </Link>
      </MenuItem>
      <MenuItem>Log In</MenuItem>
    </>
  );

  useOutsideClick(openMenuRef, () => {
    toggleMenu();
  });

  return (
    <>
      {isOpen && (
        <OpenMenu data-testid="mobile-menu" ref={openMenuRef}>
          {menuItems}
        </OpenMenu>
      )}
      <Container data-testid="container">
        <HamburgerButton data-testid="hamburger" onClick={toggleMenu}>
          <HamburgerIcon />
        </HamburgerButton>
        <Menu data-testid="menu">{menuItems}</Menu>
        <SearchBar data-testid="search" />
      </Container>
    </>
  );
};

export default Navbar;
