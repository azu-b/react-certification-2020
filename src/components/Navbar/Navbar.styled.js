import styled from 'styled-components';

const Container = styled.div`
  z-index: 1;
  width: 100%;
  position: fixed;
  top: 0;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 10px 15px;
  height: 50px;
  background-color: #b399ff;
  box-shadow: 0px 0px 13px -3px black;

  @media (min-width: 768px) {
    justify-content: space-around;
  }
`;

const SearchBar = styled.input`
  width: 200px;
  height: 30px;
  font-size: 14px;
  border: none;
  box-shadow: inset 0px 0px 11px -4px #9d80ff;

  @media (min-width: 768px) {
    width: 250px;
  }
`;

const SearchContainer = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 230px;

  @media (min-width: 768px) {
    width: 280px;
  }
`;

const Menu = styled.div`
  display: none;

  @media (min-width: 768px) {
    display: flex;
    justify-content: space-around;
    width: 250px;
  }
`;

const OpenMenu = styled.div`
  z-index: 2;
  position: fixed;
  top: 0;
  left: 0;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;
  border-right: 1px solid #c4b3ff;
  box-shadow: 0px 0px 13px -3px black;
  width: 200px;
  height: 100%;
  background-color: white;

  @media (min-width: 768px) {
    display: none;
  }
`;

const MenuItem = styled.div`
  width: 100%;
  color: #3d0099;
  font-family: 'Arial';
  font-size: 18px;
  font-weight: 500;
  padding: 10px;
  border-bottom: 1px solid #c4b3ff;

  @media (min-width: 768px) {
    color: white;
    border: none;
  }
`;

const Button = styled.div``;

const HamburgerButton = styled(Button)`
  display: flex;

  @media (min-width: 768px) {
    display: none;
  }
`;

export {
  Container,
  SearchBar,
  SearchContainer,
  Menu,
  MenuItem,
  OpenMenu,
  HamburgerButton,
};
