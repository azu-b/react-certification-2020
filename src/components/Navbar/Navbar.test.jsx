import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Navbar from './index';

describe('<Navbar>', () => {
  it('renders', () => {
    render(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>
    );
    const wrapper = screen.getByTestId('container');
    const hamburger = screen.getByTestId('hamburger');
    const menu = screen.getByTestId('menu');
    const search = screen.getByTestId('search');

    expect(wrapper.tagName).toBe('DIV');
    expect(hamburger.innerHTML).toMatch('<svg');
    expect(menu.children).toHaveLength(2);
    expect(search.innerHTML).toMatch('<input');
    expect(search.innerHTML).toMatch('<svg');
  });

  it('matches snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>
    );
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('opens the mobile menu when hamburger button is clicked', () => {
    render(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>
    );

    expect(() => {
      screen.getByTestId('mobile-menu');
    }).toThrow();

    const hamburger = screen.getByTestId('hamburger');
    fireEvent.click(hamburger);
    const mobileMenu = screen.getByTestId('mobile-menu');

    expect(mobileMenu.tagName).toBe('DIV');
    expect(mobileMenu.children).toHaveLength(2);
  });

  it('closes the mobile menu when clicked outside of it', () => {
    const onClick = jest.fn();

    render(
      <MemoryRouter>
        <button type="button" onClick={onClick}>
          Outside
        </button>
        <Navbar />
      </MemoryRouter>
    );

    const hamburger = screen.getByTestId('hamburger');
    fireEvent.click(hamburger);
    const mobileMenu = screen.getByTestId('mobile-menu');
    expect(mobileMenu.tagName).toBe('DIV');

    const button = screen.getByText('Outside');
    fireEvent.click(button);
    expect(onClick).toHaveBeenCalledTimes(1);

    expect(() => {
      screen.getByTestId('mobile-menu');
    }).toThrow();
  });
});
