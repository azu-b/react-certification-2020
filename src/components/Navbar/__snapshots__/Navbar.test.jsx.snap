// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar> matches snapshot 1`] = `
"<div data-testid=\\"container\\" class=\\"sc-hKFyIo xvRQC\\">
  <div data-testid=\\"hamburger\\" class=\\"sc-pNWxx sc-jrsJCI  jwBzmG\\"><svg viewBox=\\"0 0 28 22\\" width=\\"28\\" height=\\"22\\" fill=\\"#ffffff\\">
      <rect width=\\"28\\" height=\\"4\\" rx=\\"2\\"></rect>
      <rect y=\\"9\\" width=\\"28\\" height=\\"4\\" rx=\\"2\\"></rect>
      <rect y=\\"18\\" width=\\"28\\" height=\\"4\\" rx=\\"2\\"></rect>
    </svg></div>
  <div data-testid=\\"menu\\" class=\\"sc-gKAblj bwCMdF\\">
    <div class=\\"sc-fujyUd bikKCQ\\"><a href=\\"/\\">Home</a></div>
    <div class=\\"sc-fujyUd bikKCQ\\">Log In</div>
  </div>
  <form data-testid=\\"search\\" class=\\"sc-gtssRu iwAXyB\\"><input class=\\"sc-bdnylx sfOGj\\" value=\\"\\"><button type=\\"submit\\" class=\\"sc-dlnjPT czYKGT\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" height=\\"24px\\" viewBox=\\"-1 0 136 136.219\\" width=\\"24px\\" fill=\\"#ffffff\\">
        <path d=\\"M93.148 80.832c16.352-23.09 10.883-55.062-12.207-71.41S25.88-1.461 9.531 21.632C-6.816 44.723-1.352 76.693 21.742 93.04a51.226 51.226 0 0055.653 2.3l37.77 37.544c4.077 4.293 10.862 4.465 15.155.387 4.293-4.075 4.465-10.86.39-15.153a9.21 9.21 0 00-.39-.39zm-41.84 3.5c-18.245.004-33.038-14.777-33.05-33.023-.004-18.246 14.777-33.04 33.027-33.047 18.223-.008 33.008 14.75 33.043 32.972.031 18.25-14.742 33.067-32.996 33.098h-.023zm0 0\\"></path>
      </svg></button></form>
</div>"
`;
